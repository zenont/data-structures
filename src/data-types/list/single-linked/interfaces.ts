export type MaybeSingleLinkedListItem<T> = ISingleLinkedListItem<T> | undefined

export interface ISingleLinkedListItem<T> {
	value: T
	next?: MaybeSingleLinkedListItem<T>
}

export interface ISingleLinkedList<T> {
	pushAfter(item: ISingleLinkedListItem<T>, value: T): ISingleLinkedListItem<T>
	remove(item: ISingleLinkedListItem<T>): ISingleLinkedList<T>
	push(value: T): ISingleLinkedListItem<T>
	pop(): T | undefined
	head(): MaybeSingleLinkedListItem<T>
	tail(): MaybeSingleLinkedListItem<T>
	hasTail(): boolean
	hasHead(): boolean
}
