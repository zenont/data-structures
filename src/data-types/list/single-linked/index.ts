export * from './interfaces'
export * from './single-linked'
