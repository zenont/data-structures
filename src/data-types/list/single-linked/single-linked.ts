// import { ISingleLinkedList, ISingleLinkedListItem, MaybeSingleLinkedListItem } from './interfaces'

export type MaybeSingleLinkedListItem<T> = SingleLinkedListItem<T> | undefined
type SingleLinkedListItemType<T> = SingleLinkedListItem<T> | null

/**
 * Represents a single-linked list item.
 * THere will only be a next node reference.
 */
export class SingleLinkedListItem<T> implements SingleLinkedListItem<T> {
	public next: MaybeSingleLinkedListItem<T> = undefined
	constructor(public value: T, next?: SingleLinkedListItem<T>) {
		this.next = next || undefined
	}
}

/**
 * Single linked list.
 * This data structure is useful if you only care about iterating a sequence in forwdward direction.
 * @export
 * @class SingleLinkedList
 * @implements {ISingleLinkedList<T>}
 * @template T
 */
export class SingleLinkedList<T> implements SingleLinkedList<T> {
	/**
	 * Creates a single linked item from a set of values.
	 * @static
	 * @template T
	 * @param {...T[]} values
	 * @returns {ISingleLinkedList<T>}
	 * @memberof SingleLinkedList
	 */
	public static of<T>(...values: T[]): SingleLinkedList<T> {
		return new SingleLinkedList<T>(...values)
	}

	private _head: MaybeSingleLinkedListItem<T> = undefined
	private _tail: MaybeSingleLinkedListItem<T> = undefined

	constructor(...values: T[]) {
		values.forEach(item => this.push(item))
	}

	/**
	 * Pushes a value node after the referenced node.
	 * @param {ISingleLinkedListItem<T>} item
	 * @param {T} value
	 * @returns {ISingleLinkedListItem<T>}
	 * @memberof SingleLinkedList
	 */
	public pushAfter(item: SingleLinkedListItem<T>, value: T): SingleLinkedListItem<T> {
		const itemToPush = new SingleLinkedListItem(value)
		if (item.next != null) {
			itemToPush.next = item.next
		}
		item.next = itemToPush
		if (item === this._tail) {
			this._tail = itemToPush
		}
		return itemToPush
	}

	/**
	 * Removes a node from the list.
	 * Both, head & tail references will be updated if affected.
	 * @param {SingleLinkedListItem<T>} item
	 * @returns {SingleLinkedList<T>}
	 * @memberof SingleLinkedList
	 */
	public remove(item: SingleLinkedListItem<T>): SingleLinkedList<T> {
		if (item == null || this._head == null) return this

		if (this._head === item) {
			this._head = item.next
			if (item.next == null) {
				this._tail = undefined
			}
			return this
		}

		let prev: MaybeSingleLinkedListItem<T> = this._head
		if (prev == null) return this

		let next = prev.next
		if (next == null) return this

		while (next !== item) {
			prev = next
			next = next != null ? next.next : undefined
		}
		if (prev != null) {
			prev.next = next.next
			if (prev.next == null) {
				// new tail!
				this._tail = prev
			}
		}
		return this
	}

	/**
	 * Pushes a value to the list as a node.
	 * @param {T} value
	 * @returns {SingleLinkedListItem<T>}
	 * @memberof SingleLinkedList
	 */
	public push(value: T): SingleLinkedListItem<T> {
		const item = new SingleLinkedListItem(value)
		if (this._head == null) {
			this._head = item
			this._tail = item
		} else {
			const itemToReplace = this._tail
			itemToReplace!.next = item
			this._tail = item
		}
		return item
	}

	/**
	 * It pops a value out of the list.
	 * The node with the value will be removed.
	 * @returns {(T | undefined)}
	 * @memberof SingleLinkedList
	 */
	public pop(): T | undefined {
		const tail = this.tail()
		if (tail == null) return undefined

		this.remove(tail)
		return tail.value
	}

	/**
	 * Gets the head node.
	 * @returns {MaybeSingleLinkedListItem<T>}
	 * @memberof SingleLinkedList
	 */
	public head(): MaybeSingleLinkedListItem<T> {
		return this._head
	}

	/**
	 * Gets the tail node.
	 * @returns {MaybeSingleLinkedListItem<T>}
	 * @memberof SingleLinkedList
	 */
	public tail(): MaybeSingleLinkedListItem<T> {
		return this._tail
	}

	/**
	 * Returns true if the tail node exists, false otherwise.
	 * @returns {boolean}
	 * @memberof SingleLinkedList
	 */
	public hasTail(): boolean {
		return this._tail != null
	}

	/**
	 * Returns true if the head node exists, false otherwise.
	 * @returns {boolean}
	 * @memberof SingleLinkedList
	 */
	public hasHead(): boolean {
		return this._head != null
	}

	private findFromLast(index: number): MaybeSingleLinkedListItem<T> {
		if (this._head == null) return undefined
		if (index === 0) return this._tail

		let candidate: SingleLinkedListItemType<T> = this._head
		let current: SingleLinkedListItemType<T> = this._head
		for (let i = 0; i < index; i++) {
			if (current.next) {
				current = current.next
			} else {
				return undefined
			}
		}
		while (current.next) {
			current = current.next
			if (candidate && candidate.next) {
				candidate = candidate.next
			}
		}

		return candidate
	}
}
