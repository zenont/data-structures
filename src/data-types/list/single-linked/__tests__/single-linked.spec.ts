import { ISingleLinkedList, ISingleLinkedListItem } from '../interfaces'
import { SingleLinkedList } from '../single-linked'

describe('single linked List', () => {
	it('should create', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)

		expect(linkedList).toMatchSnapshot()
	})

	it('should create empty', () => {
		const linkedList = SingleLinkedList.of()

		expect(linkedList).toMatchSnapshot()
	})

	it('should push one item', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)
		linkedList.push(2)

		expect(linkedList).toMatchSnapshot()
	})

	it('should push many items', () => {
		const linkedList = SingleLinkedList.of()
		linkedList.push(1)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)

		expect(linkedList).toMatchSnapshot()
	})

	it('should push many items with seed', () => {
		const linkedList = SingleLinkedList.of(1)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)

		expect(linkedList).toMatchSnapshot()
	})

	it('should push after item', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)
		const item = linkedList.push(2)
		linkedList.push(4)
		linkedList.pushAfter(item, 3)
		expect(linkedList).toMatchSnapshot()
	})

	it('should push after many items', () => {
		const linkedList = SingleLinkedList.of()
		let item = linkedList.push(1)
		item = linkedList.push(2)
		item = linkedList.pushAfter(item, 3)
		item = linkedList.pushAfter(item, 4)
		item = linkedList.pushAfter(item, 5)
		expect(linkedList).toMatchSnapshot()
	})

	it('should remove item in the middle', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		const toRemove = linkedList.tail()
		linkedList.push(4)
		linkedList.push(5)

		linkedList.remove(toRemove)
		expect(linkedList).toMatchSnapshot()
	})

	it('should remove head', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		const toRemove = linkedList.head()

		linkedList.remove(toRemove!)
		expect(linkedList).toMatchSnapshot()
	})

	it('should remove tail', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		const toRemove = linkedList.tail()

		linkedList.remove(toRemove!)
		expect(linkedList).toMatchSnapshot()
	})

	it('should remove all items', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		let toRemove = linkedList.tail()
		linkedList.remove(toRemove!)
		toRemove = linkedList.tail()
		linkedList.remove(toRemove!)
		toRemove = linkedList.tail()
		linkedList.remove(toRemove!)
		expect(linkedList).toMatchSnapshot()
	})

	it('should pop item', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		const popped = linkedList.pop()
		expect(popped).toMatchSnapshot()
		expect(linkedList).toMatchSnapshot()
	})

	it('should pop all items', () => {
		const seed = 1
		const linkedList = SingleLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		let popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		expect(popped).toMatchSnapshot()
		expect(linkedList).toMatchSnapshot()
	})
	/*
		it('should find from last index at tail', () => {
			const seed = 1
			const linkedList = SingleLinkedList.of(seed)
			linkedList.add(2)
			linkedList.add(3)
			linkedList.add(4)
			linkedList.add(5)
			const found = linkedList.findFromLast(0)
			expect(found).toMatchSnapshot()
		})

		it('should find from last index at 0', () => {
			const seed = 1
			const linkedList = SingleLinkedList.of(seed)
			linkedList.add(2)
			linkedList.add(3)
			linkedList.add(4)
			linkedList.add(5)
			const found = linkedList.findFromLast(0)
			expect(found).toMatchSnapshot()
		})

		it('should find from last index at 1', () => {
			const seed = 1
			const linkedList = SingleLinkedList.of(seed)
			linkedList.add(2)
			linkedList.add(3)
			linkedList.add(4)
			linkedList.add(5)
			const found = linkedList.findFromLast(1)
			expect(found).toMatchSnapshot()
		})

		it('should find from last index at 2', () => {
			const seed = 1
			const linkedList = SingleLinkedList.of(seed)
			linkedList.add(2)
			linkedList.add(3)
			linkedList.add(4)
			linkedList.add(5)
			const found = linkedList.findFromLast(2)
			expect(found).toMatchSnapshot()
		})

		it('should find from last index at 3', () => {
			const seed = 1
			const linkedList = SingleLinkedList.of(seed)
			linkedList.add(2)
			linkedList.add(3)
			linkedList.add(4)
			linkedList.add(5)
			const found = linkedList.findFromLast(3)
			expect(found).toMatchSnapshot()
		})

		it('should find from last index at 4', () => {
			const seed = 1
			const linkedList = SingleLinkedList.of(seed)
			linkedList.add(2)
			linkedList.add(3)
			linkedList.add(4)
			linkedList.add(5)
			const found = linkedList.findFromLast(4)
			expect(found).toMatchSnapshot()
		})

		it('should find from last index at 5', () => {
			const seed = 1
			const linkedList = SingleLinkedList.of(seed)
			linkedList.add(2)
			linkedList.add(3)
			linkedList.add(4)
			linkedList.add(5)
			const found = linkedList.findFromLast(5)
			expect(found).toMatchSnapshot()
		})

		it('should find from last index at 5', () => {
			const seed = 1
			const linkedList = SingleLinkedList.of(seed)
			linkedList.add(2)
			linkedList.add(3)
			linkedList.add(4)
			linkedList.add(5)
			const found = linkedList.findFromLast(5)
			expect(found).toMatchSnapshot()
		})

		it('should create list from array', () => {
			const linkedList = SingleLinkedList.of(1, 2, 3, 4, 5)
			expect(linkedList).toMatchSnapshot()
		})

		it('should fluent add', () => {
			const linkedList = SingleLinkedList.of(1).add(2).add(3).add(4).add(5)
			expect(linkedList).toMatchSnapshot()
		})

		it('should fluent add and remove', () => {
			const linkedList = SingleLinkedList.of(1).add(2).add(3)
			const toRemove = linkedList.tail()
			linkedList.add(4).add(5).add(6)
			linkedList.remove(toRemove)
			expect(linkedList).toMatchSnapshot()
		})

		it('should pop', () => {
			const linkedList = SingleLinkedList.of(1).add(2).add(3)
			linkedList.pop()
			expect(linkedList).toMatchSnapshot()
		})

		it('should over pop', () => {
			const linkedList = SingleLinkedList.of(1).add(2).add(3)
			linkedList.pop()
			linkedList.pop()
			linkedList.pop()
			linkedList.pop()
			expect(linkedList).toMatchSnapshot()
		})*/
})
