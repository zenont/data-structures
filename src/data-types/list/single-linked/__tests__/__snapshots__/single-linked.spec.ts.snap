// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single linked List should create 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": undefined,
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 1,
  },
}
`;

exports[`single linked List should create empty 1`] = `
SingleLinkedList {
  "_head": undefined,
  "_tail": undefined,
}
`;

exports[`single linked List should create from array 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": SingleLinkedListItem {
            "next": undefined,
            "value": 5,
          },
          "value": 4,
        },
        "value": 3,
      },
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 5,
  },
}
`;

exports[`single linked List should over pop 1`] = `undefined`;

exports[`single linked List should over pop 2`] = `
SingleLinkedList {
  "_head": undefined,
  "_tail": undefined,
}
`;

exports[`single linked List should pop all items 1`] = `1`;

exports[`single linked List should pop all items 2`] = `
SingleLinkedList {
  "_head": undefined,
  "_tail": undefined,
}
`;

exports[`single linked List should pop item 1`] = `5`;

exports[`single linked List should pop item 2`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": undefined,
          "value": 4,
        },
        "value": 3,
      },
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 4,
  },
}
`;

exports[`single linked List should push after item 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": undefined,
          "value": 4,
        },
        "value": 3,
      },
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 4,
  },
}
`;

exports[`single linked List should push after many items 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": SingleLinkedListItem {
            "next": undefined,
            "value": 5,
          },
          "value": 4,
        },
        "value": 3,
      },
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 5,
  },
}
`;

exports[`single linked List should push many items 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": SingleLinkedListItem {
            "next": undefined,
            "value": 5,
          },
          "value": 4,
        },
        "value": 3,
      },
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 5,
  },
}
`;

exports[`single linked List should push many items with seed 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": SingleLinkedListItem {
            "next": undefined,
            "value": 5,
          },
          "value": 4,
        },
        "value": 3,
      },
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 5,
  },
}
`;

exports[`single linked List should push one item 1`] = `
SingleLinkedListItem {
  "next": undefined,
  "value": 2,
}
`;

exports[`single linked List should push one item 2`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": undefined,
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 2,
  },
}
`;

exports[`single linked List should remove all items 1`] = `
SingleLinkedList {
  "_head": undefined,
  "_tail": undefined,
}
`;

exports[`single linked List should remove head 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": undefined,
          "value": 5,
        },
        "value": 4,
      },
      "value": 3,
    },
    "value": 2,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 5,
  },
}
`;

exports[`single linked List should remove item in the middle 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": undefined,
          "value": 5,
        },
        "value": 4,
      },
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 5,
  },
}
`;

exports[`single linked List should remove tail 1`] = `
SingleLinkedList {
  "_head": SingleLinkedListItem {
    "next": SingleLinkedListItem {
      "next": SingleLinkedListItem {
        "next": SingleLinkedListItem {
          "next": undefined,
          "value": 4,
        },
        "value": 3,
      },
      "value": 2,
    },
    "value": 1,
  },
  "_tail": SingleLinkedListItem {
    "next": undefined,
    "value": 4,
  },
}
`;
